draft-ietf-tls-esni.md

```diff
@@ -446,33 +482,148 @@ to harmonize these to make sure that we maintain key separation.]]
+It then SHOULD retry the handshake with a new transport connection and encrypted
+SNI disabled.
+
+[[TODO: Key replacement is significantly less scary than saying that ESNI-naive
+  servers bounce ESNI off. Is it worth defining a strict mode toggle in the ESNI
+  keys, for a deployment to indicate it is ready for that? ]]
+
+The client MUST NOT treat this as a signal to disable encrypted SNI until the
+handshake completes successfully.
+
+Clients SHOULD implement a limit on retries caused by "esni_retry_request" or
+servers which do not acknowledge the "encrypted_server_name" extension. If the
+client does not retry in either scenario, it MUST report an error to the
+calling application.
+
+### Verifying against the public name {#verify-public-name}
+
+When the server cannot decrypt or does not process the "encrypted_server_name"
+extension, it continues with the handshake using the cleartext "server_name"
+extension instead (see {{server-behavior}}). Clients that offer ESNI then
+verify the certificate with the public name, as follows:
+
+- If the server resumed a session or negotiated parameters not based on
+  certificates, the client MUST abort the connection with an illegal_parameter
+  alert. This case is invalid because {{client-behavior}} requires the client
+  to only offer ESNI-established sessions, and {{server-behavior}} requires
+  the server to decline ESNI-established sessions if it did not accept ESNI.
+
+- The client MUST verify that the certificate is valid for ESNIKeys.public_name.
+  If invalid, it MUST abort the connection with the appropriate alert.
+
+- If the server requests a client certificate, the client MUST respond with an
+  empty Certificate message, denoting no client certificate.
+
+Note that verifying a connection for the public name does not verify it for the
+origin. The TLS implementation MUST NOT surface such connections as successful to
+the application. It additionally MUST ignore all session tickets and session IDs
+presented by the server. These connections are only used to trigger retries, as
+described in {{client-behavior}}. This may be implemented, for instance, by
+reporting a failed connection with a dedicated error code.
+
+### GREASE extensions {#grease-extensions}
+
+If the client attempts to connect to a server and does not have an ESNIKeys
+structure available for the server, it SHOULD send a GREASE
+{{I-D.ietf-tls-grease}} "encrypted_server_name" extension as follows:
+
+- Select a supported cipher suite, named group, and padded_length
+  value. The padded_length value SHOULD be 260. Set the "suite" field
+  to the selected cipher suite.
+
+- Set the "key_share" field to a randomly-generated valid public key
+  for the named group.
+
+- Set the "record_digest" field to a randomly-generated string of hash_length
+  bytes, where hash_length is the length of the hash function associated with
+  the chosen cipher suite.
```

Comment:
Should we instead use consistent hashing so that you had the same record_digest for each server.